Prevent alerts to be captured

maxbritto opened the pull request:
- Calls to the snapshot() function will now wait until all alert have
  disappeared before taking the screenshot (can be disabled with a
  boolean parameter on the snapshot() method)
- Adds a function to allow automatic dismissal of alerts that prevent
  actions or snapshots
- Copy the new SnapshotAlertManagement.swift file to the same place as
  the SnapshotHelper.swift file during original setup

maxbritto committed:
Prevent alerts to be captured

- Calls to the snapshot() function will now wait until all alert have
disappeared before taking the screenshot (can be disabled with a
boolean parameter on the snapshot() method)
- Adds a function to allow automatic dismissal of alerts that prevent
actions or snapshots
- Copy the new SnapshotAlertManagement.swift file to the same place as
the SnapshotHelper.swift file during original setup

koenpunt commented:
Is the `waitForAlerts` argument necessary? Shouldn't it be that if `automaticallyDismissAlerts` is called, all alert are dismissed? And if someone doesn't want an alert to be dismissed, they simply call `stopDismissingAlerts`

koenpunt commented:
Or if it stays, I think it should be named differently, because now it implies to wait for an alert to be shown, instead of the dismissal thereof

maxbritto commented:
I'm not sure because `waitForAlerts` actually does wait but doesn't dismiss the alert by itself. It does the same thing as `waitForLoadingIndicator`. It lets the developer chose its own way of dismissing the alert : he can take care of it on it's own if the alert matters or call `automaticallyDismissAlerts` to automatically dismiss any alert that gets in the way.

maxbritto commented:
Regarding your second comment, I see what you mean but then `waitForLoadingIndicator` should also be renamed.

koenpunt commented:
> Regarding your second comment, I see what you mean but then waitForLoadingIndicator should also be renamed.

You're right.

> I'm not sure because waitForAlerts actually does wait but doesn't dismiss the alert by itself. It does the same thing as waitForLoadingIndicator. 

`waitForLoading` indicator waits for the loading indicator to disappear, but you say that `waitForAlerts` doesn't dismiss alert? That might be confusing to users

maxbritto commented:
I tried to follow the pattern of waitForLoading and do the same : I just wait for the alert to disappear. This is based on the assumption that many developer are probably doing meaningful things with their alerts and do not want us to dismiss them as they appear.
But I've also added a print if we have been waiting for more than 2 seconds and an alert is still here : it explains how to enable the automatic dismissal if they want it. I also added documentation in setup.rb

makadaw commented on lib/assets/SnapshotAlertManager.swift:
```diff
@@ -0,0 +1,60 @@
+              if alertInterruptionMonitor == nil {
+                  print("The SnapshotAlertManager can automatically hide alerts that will prevent actions or screenshots. Just call this method to enable it : automaticallyDismissAlerts(testCase:XCTestCase)")
+              } else {
+                  XCUIApplication().tap() //UIInterruptionMonitor will only trigger if an action is prevented by the alert so we generate one if the user script hasn't after 2 seconds
```
XCUIApplication().tap() may tap on some control if app do not display alert

maxbritto replied:
We arrive here only after an alert has been present for at least 2 seconds so the app will be displaying the alert

maxbritto committed:
Merge remote-tracking branch 'fastlane/master' into alert-management

* fastlane/master:
  Updated launch_arguments to be an array
  Bump version
  Improved multiprocess support.
  Update the update check for the embedded SnapshotHelper.swift
  Change KrauseFx to fastlane in the title of the summary template
  Update README.MD
  Trim whitespace & newlines in locale string
  Removed class variable
  Version bump to 1.9.0
  Update the updater code to check for the newest SwiftHelper version
  Fixed `reset_simulators` command.

i2amsam commented:
Hey @maxbritto thank you for this!  

I'm not going to merge this for now, though I think this an interesting problem, I'm concerned about adding additional dependencies and the migration path for existing setups. This could be a good candidate for getting "local helpers" or "local action" for some sort of plugin system in `snapshot` so the helpers can be easily extended.

maxbritto commented:
I've never seen any plugin system for fastlane/snapshot. If there is one I'd be willing to refactor this as a plug-in. But if you're asking me to help building the plugin system I'm not sure how efficient I would be since I've never done any Ruby.
